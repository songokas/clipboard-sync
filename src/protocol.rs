use core::str::FromStr;

use clap::ValueEnum;
use serde::{Deserialize, Serialize};

use crate::errors::CliError;

#[derive(
    Default,
    ValueEnum,
    Serialize,
    Deserialize,
    Debug,
    Clone,
    PartialEq,
    Eq,
    Hash,
    PartialOrd,
    Ord,
    Copy,
)]
#[repr(u8)]
#[serde(rename_all = "kebab-case")]
pub enum Protocol {
    #[default]
    Basic,
    Tcp,
    TcpTls,
    Quic,
}

impl TryFrom<u8> for Protocol {
    type Error = ();

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        Ok(match value {
            0 => Self::Basic,
            1 => Self::Tcp,
            2 => Self::TcpTls,
            3 => Self::Quic,
            _ => return Err(()),
        })
    }
}

impl FromStr for Protocol {
    type Err = CliError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match s {
            "basic" => Protocol::Basic,
            "tcp" => Protocol::Tcp,
            "tcp-tls" => Protocol::TcpTls,
            "quic" => Protocol::Quic,
            v => {
                return Err(CliError::ArgumentError(format!(
                    "Protocol {} is not available",
                    v
                )));
            }
        })
    }
}

impl Protocol {
    #[allow(unused_variables)]
    pub fn from(protocol_opt: Option<&str>) -> Result<Protocol, CliError> {
        let protocol = match protocol_opt {
            Some(s) => s.parse()?,
            None => Protocol::Basic,
        };
        Ok(protocol)
    }
}

impl std::fmt::Display for Protocol {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Quic => write!(f, "quic"),
            Self::Basic => write!(f, "basic"),
            Self::Tcp => write!(f, "tcp"),
            Self::TcpTls => write!(f, "tcp-tls"),
        }
    }
}
