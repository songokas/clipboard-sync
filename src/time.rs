use chrono::Utc;

#[cfg(feature = "ntp")]
use crate::errors::ConnectionError;
use crate::errors::ValidationError;
#[cfg(feature = "ntp")]
use log::{debug, warn};
#[cfg(feature = "ntp")]
use rsntp::AsyncSntpClient;
#[cfg(feature = "ntp")]
use std::sync::atomic::AtomicI64;

#[cfg(feature = "ntp")]
use std::sync::atomic::Ordering;
use std::time::Duration;

#[cfg(feature = "ntp")]
static TIME_DIFF: AtomicI64 = AtomicI64::new(0);

pub fn get_time() -> u64 {
    #[cfg(feature = "ntp")]
    {
        let diff = TIME_DIFF.load(Ordering::Relaxed);
        let now = Utc::now().timestamp();
        (now + diff) as u64
    }
    #[cfg(not(feature = "ntp"))]
    return Utc::now().timestamp() as u64;
}

pub fn validate_timestamp(
    now: u64,
    timestamp: u64,
    max_expected: Duration,
) -> Result<(), ValidationError> {
    // let now = get_time();
    let difference = if now >= timestamp {
        now - timestamp
    } else {
        timestamp - now
    };
    if max_expected.as_secs() >= difference {
        Ok(())
    } else {
        Err(ValidationError::InvalidTimestamp {
            difference,
            max_expected,
        })
    }
}

#[cfg(feature = "ntp")]
pub async fn update_time_diff(
    ntp_address: String,
    cancel: tokio_util::sync::CancellationToken,
) -> crate::defaults::ExecutorResult {
    use tokio::{
        select,
        time::{self},
    };

    let mut interval = time::interval(Duration::from_secs(1800));
    let mut success_count = 0;
    loop {
        select! {
            _ = cancel.cancelled() => {
                debug!("Ntp cancelled");
                break;
            }
            _ = interval.tick() => {
                let diff = get_time_diff(&ntp_address).await;
                TIME_DIFF.fetch_add(diff, Ordering::Relaxed);
                success_count += 1;
            }
        }
    }
    Ok(("ntp", success_count))
}

#[cfg(feature = "ntp")]
async fn get_time_diff(ntp_address: &str) -> i64 {
    let real = if let Ok(t) = get_ntp_time(ntp_address).await {
        t as i64
    } else {
        return 0;
    };
    let now = Utc::now().timestamp();
    if real >= now {
        real - now
    } else {
        now - real
    }
}

#[cfg(feature = "ntp")]
async fn get_ntp_time(ntp_address: &str) -> Result<u64, ConnectionError> {
    let client = AsyncSntpClient::new();
    let result = client.synchronize(ntp_address).await.map_err(|e| {
        warn!("Failed to synchronize time: {}", e);
        ConnectionError::FailedToConnect(format!(
            "Failed to retrieve time from ntp server {} {}",
            ntp_address, e
        ))
    })?;
    let seconds = result
        .datetime()
        .unix_timestamp()
        .expect("ntp unix timestamp");
    debug!(
        "ntp time updated. server: {} utc time in seconds: {} utc local time: {}",
        ntp_address,
        seconds.as_secs(),
        Utc::now().timestamp()
    );
    Ok(seconds.as_secs())
}

#[cfg(test)]
mod tests {
    use test_data_file::test_data_file;

    use super::*;

    #[test]
    fn test_get_time() {
        assert!(get_time() > 0);
    }

    #[test_data_file(path = "tests/samples/validate_timestamp.csv")]
    #[test]
    fn test_validate_timestamp(now: u64, timestamp: u64, max_valid_secs: u64, is_ok: bool) {
        assert_eq!(
            validate_timestamp(now, timestamp, Duration::from_secs(max_valid_secs)).is_ok(),
            is_ok
        );
    }
}



