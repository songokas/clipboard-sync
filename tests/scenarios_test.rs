use clipboard_sync::protocol::Protocol;
use insta_cmd::assert_cmd_snapshot;
use rand::{distributions::Alphanumeric, Rng};
use std::fs::{create_dir_all, remove_dir, remove_file, write};
use std::thread;
use std::time::Duration;
use test_data_file::test_data_file;

mod common;

use crate::common::run_command;

const ANY_KEY: &str = "12345678912345678912345678912345";

#[test]
#[ignore]
//cargo test test_receive_once_multicast -- --ignored
fn test_receive_once_multicast() {
    assert_cmd_snapshot!(run_command(
        vec![
            "--key",
            ANY_KEY,
            "--receive-once",
            "--receive-once-wait",
            "1",
            "--verbosity",
            "debug",
        ],
        Duration::from_secs(10),
        "",
        true,
        vec![],
    ));
}

#[test]
#[ignore]
// cargo test test_send_once_multicast -- --ignored
fn test_send_once_multicast() {
    assert_cmd_snapshot!(run_command(
        vec!["--key", ANY_KEY, "--send-once", "--clipboard", "/dev/stdin"],
        Duration::from_secs(2),
        "hello",
        true,
        vec![],
    ));
}

#[test]
fn test_send_heartbeat_without_sending_clipboard_data() {
    assert_cmd_snapshot!(run_command(
        vec![
            "--key",
            ANY_KEY,
            "--clipboard",
            "/dev/stdin",
            "--ignore-initial-clipboard",
            "--heartbeat",
            "10",
            "--verbosity",
            "debug=simple",
            "--bind-address",
            "0.0.0.0:33321",
        ],
        Duration::from_secs(3),
        "hello",
        true,
        vec![],
    ));
}

#[test]
fn test_file_changes() {
    let file = "/tmp/test_file_changes";
    let _ = remove_file(file);
    write(file, b"testdata1").unwrap();
    let t1 = thread::spawn(move || {
        assert_cmd_snapshot!(run_command(
            vec![
                "--key",
                ANY_KEY,
                "--clipboard",
                file,
                "--ignore-initial-clipboard",
                "--verbosity",
                "debug=simple",
                "--bind-address",
                "0.0.0.0:33322",
            ],
            Duration::from_secs(3),
            "hello",
            true,
            vec![],
        ))
    });
    thread::sleep(Duration::from_millis(1000));
    write(file, b"testdata2").unwrap();
    t1.join().unwrap();
    remove_file(file).unwrap();
}

#[test]
fn test_file_changes_created_after_startup() {
    let dir = "/tmp/_test_file_changes_created_after_startup";
    create_dir_all(dir).unwrap();
    let file = format!("{dir}/file");
    remove_file(&file).unwrap_or(());
    let cfile = file.clone();
    let t1 = thread::spawn(move || {
        assert_cmd_snapshot!(run_command(
            vec![
                "--key",
                ANY_KEY,
                "--clipboard",
                &cfile,
                "--ignore-initial-clipboard",
                "--verbosity",
                "debug=simple",
                "--bind-address",
                "0.0.0.0:33323",
            ],
            Duration::from_secs(3),
            "hello",
            true,
            vec![],
        ))
    });
    thread::sleep(Duration::from_millis(1000));
    write(&file, b"testdata2").unwrap();
    t1.join().unwrap();
    remove_file(file).unwrap();
}

#[test]
fn test_directory_changes() {
    let dir = "/tmp/_test_directory_changes";
    create_dir_all(dir).unwrap();
    let file = format!("{}/_random_file", dir);
    write(&file, b"testdata1").unwrap();
    let t1 = thread::spawn(move || {
        assert_cmd_snapshot!(run_command(
            vec![
                "--key",
                ANY_KEY,
                "--clipboard",
                dir,
                "--ignore-initial-clipboard",
                "--verbosity",
                "debug=simple",
                "--bind-address",
                "0.0.0.0:33324",
            ],
            Duration::from_secs(3),
            "hello",
            true,
            vec![],
        ))
    });
    thread::sleep(Duration::from_millis(1000));
    write(&file, b"testdata2").unwrap();
    t1.join().unwrap();
    remove_file(&file).unwrap();
    remove_dir(dir).unwrap();
}

#[test]
fn test_directory_changes_created_after_startup() {
    let dir = "/tmp/_test_directory_changes_created_after_startup/dir";
    create_dir_all(dir).unwrap();
    let file = format!("{}/file", dir);
    remove_file(&file).unwrap_or(());
    remove_dir(dir).unwrap_or(());
    let t1 = thread::spawn(move || {
        assert_cmd_snapshot!(run_command(
            vec![
                "--key",
                ANY_KEY,
                "--clipboard",
                dir,
                "--ignore-initial-clipboard",
                "--verbosity",
                "debug=simple",
                "--bind-address",
                "0.0.0.0:33325",
            ],
            Duration::from_secs(6),
            "hello",
            true,
            vec![],
        ))
    });
    thread::sleep(Duration::from_millis(1000));
    create_dir_all(dir).unwrap();
    // there is a delay between listen for directory and writing to file in that directory
    thread::sleep(Duration::from_millis(2000));
    write(&file, b"testdata1").unwrap();
    thread::sleep(Duration::from_millis(1000));
    t1.join().unwrap();
    remove_file(&file).unwrap();
    remove_dir(dir).unwrap();
}

#[test_data_file(path = "tests/samples/send_receive_same_port.list")]
#[test]
fn test_send_receive_same_port(protocol: Protocol) {
    let size = 2000;

    let bind_to = "127.0.0.1:12910";
    let allowed_host = "127.0.0.1:12911=localhost";

    let t1 = thread::spawn(move || {
        assert_cmd_snapshot!(
            format!("test_send_receive_same_port_client1_{protocol}"),
            run_command(
                vec![
                    "--verbosity",
                    "debug=simple",
                    "--key",
                    ANY_KEY,
                    "--bind-address",
                    bind_to,
                    "--send-using-address",
                    bind_to,
                    "--allowed-host",
                    allowed_host,
                    "--protocol",
                    &protocol.to_string(),
                    "--clipboard",
                    "/dev/stdin",
                    "--remote-certificates",
                    "tests/certs/cert-verify/for-client",
                    "--private-key",
                    "tests/certs/testclient.key",
                    "--certificate-chain",
                    "tests/certs/testclient.crt",
                ],
                Duration::from_secs(3),
                "client that sends",
                true,
                vec![
                    "drive; id".to_string(),
                    "TLS1.3 encrypted extensions".to_string(),
                    "Got CertificateRequest CertificateRequestPayloadTls13".to_string(),
                ],
            )
        )
    });

    let contents: String = rand::thread_rng()
        .sample_iter(&Alphanumeric)
        .take(size)
        .map(char::from)
        .collect();

    thread::sleep(Duration::from_millis(100));

    let allowed_host = "127.0.0.1:12910=testclient";
    let bind_to = "127.0.0.1:12911";

    let t2 = thread::spawn(move || {
        assert_cmd_snapshot!(
            format!("test_send_receive_same_port_client2_{protocol}"),
            run_command(
                vec![
                    "--verbosity",
                    "debug=simple",
                    "--key",
                    ANY_KEY,
                    "--bind-address",
                    bind_to,
                    "--send-using-address",
                    bind_to,
                    "--clipboard",
                    "/dev/stdin",
                    "--allowed-host",
                    allowed_host,
                    "--protocol",
                    &protocol.to_string(),
                    "--remote-certificates",
                    "tests/certs/cert-verify/for-server",
                    "--private-key",
                    "tests/certs/localhost.key",
                    "--certificate-chain",
                    "tests/certs/localhost.crt",
                ],
                Duration::from_secs(3),
                &contents,
                true,
                vec![
                    "drive; id".to_string(),
                    "TLS1.3 encrypted extensions".to_string(),
                    "Got CertificateRequest CertificateRequestPayloadTls13".to_string(),
                ],
            )
        )
    });

    t1.join().unwrap();
    t2.join().unwrap();
}

#[test_data_file(path = "tests/samples/send_receive_once.list")]
#[test]
fn test_send_receive_once(protocol: Protocol, size: usize) {
    let bind_to = "127.0.0.1:12903,[::1]:12903";
    let allowed_host = "127.0.0.1:0,[::1]:0";

    let t1 = thread::spawn(move || {
        assert_cmd_snapshot!(
            format!("test_send_receive_once_server_{protocol}_{size}"),
            run_command(
                vec![
                    "--verbosity",
                    "debug=simple",
                    "--key",
                    ANY_KEY,
                    "--bind-address",
                    bind_to,
                    "--receive-once",
                    "--allowed-host",
                    allowed_host,
                    "--receive-once-wait",
                    "1",
                    "--protocol",
                    &protocol.to_string(),
                    "--remote-certificates",
                    "tests/certs/cert-verify/for-server",
                    "--private-key",
                    "tests/certs/localhost.key",
                    "--certificate-chain",
                    "tests/certs/localhost.crt",
                ],
                Duration::from_secs(10),
                "",
                true,
                vec![
                    "drive; id".to_string(),
                    "TLS1.3 encrypted extensions".to_string(),
                    "Got CertificateRequest CertificateRequestPayloadTls13".to_string(),
                ],
            )
        )
    });

    let contents: String = rand::thread_rng()
        .sample_iter(&Alphanumeric)
        .take(size)
        .map(char::from)
        .collect();

    let send_to = "127.0.0.1:12903=localhost";

    thread::sleep(Duration::from_millis(100));

    let t2 = thread::spawn(move || {
        assert_cmd_snapshot!(
            format!("test_send_receive_once_client_{protocol}_{size}"),
            run_command(
                vec![
                    "--verbosity",
                    "debug=simple",
                    "--key",
                    ANY_KEY,
                    "--send-using-address",
                    "127.0.0.1:12904,[::1]:12904",
                    "--send-once",
                    "--clipboard",
                    "/dev/stdin",
                    "--allowed-host",
                    send_to,
                    "--protocol",
                    &protocol.to_string(),
                    "--remote-certificates",
                    "tests/certs/cert-verify/for-client",
                    "--private-key",
                    "tests/certs/testclient.key",
                    "--certificate-chain",
                    "tests/certs/testclient.crt",
                ],
                Duration::from_secs(4),
                &contents,
                true,
                vec![
                    "drive; id".to_string(),
                    "TLS1.3 encrypted extensions".to_string(),
                    "Got CertificateRequest CertificateRequestPayloadTls13".to_string(),
                ],
            )
        )
    });

    t1.join().unwrap();
    t2.join().unwrap();
}
