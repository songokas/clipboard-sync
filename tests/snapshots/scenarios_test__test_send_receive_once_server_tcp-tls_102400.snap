---
source: tests/scenarios_test.rs
description: "Ntp server not provided\nStarting clipboard writer\nWaiting to receive clipboard once\nLoading certificates from configuration directory\nLoading certificates from configuration directory\nListening on local_addr=127.0.0.1:12903 protocol=tcp-tls multicast_addr=None certificate_serial=46:8d:f7:de:e1:27:f9:0d:06:c6:1f:ea:05:36:d8:2e:ad:69:09:35 certificate_dns=localhost\nListening on local_addr=[::1]:12903 protocol=tcp-tls multicast_addr=None certificate_serial=46:8d:f7:de:e1:27:f9:0d:06:c6:1f:ea:05:36:d8:2e:ad:69:09:35 certificate_dns=localhost\nNew connection remote_addr=127.0.0.1:12904\ndecided upon suite TLS13_AES_256_GCM_SHA384\nWaiting for stream from remote_addr=127.0.0.1:12904\nTcp receive stream local_addr=127.0.0.1:12903 remote_addr=127.0.0.1:12904\nReceived stream expected_size=102444\nValidate data identity=127.0.0.1\nMessage received data_size=102400 group=default message_type=file\nWaiting for stream from remote_addr=127.0.0.1:12904\nTcp receive stream local_addr=127.0.0.1:12903 remote_addr=127.0.0.1:12904\nReceiver connection closed: Connection expected data, but none was received\nSending warning alert CloseNotify\nReceived bytes 102400 in clipboard\nCancel reader\nCancel reader\nFinished tcptls reader processed messages 0\nFinished tcptls reader processed messages 1\nFinished protocol to clipboard forwarder processed messages 1\nFinished clipboard writer processed messages 1\n"
info:
  program: clipboard-sync
  args:
    - "--verbosity"
    - debug=simple
    - "--key"
    - "12345678912345678912345678912345"
    - "--bind-address"
    - "127.0.0.1:12903,[::1]:12903"
    - "--receive-once"
    - "--allowed-host"
    - "127.0.0.1:0,[::1]:0"
    - "--receive-once-wait"
    - "1"
    - "--protocol"
    - tcp-tls
    - "--remote-certificates"
    - tests/certs/cert-verify/for-server
    - "--private-key"
    - tests/certs/localhost.key
    - "--certificate-chain"
    - tests/certs/localhost.crt
snapshot_kind: text
---
success: true
exit_code: 0
----- stdout -----

----- stderr -----
Cancel reader
Cancel reader
Finished clipboard writer processed messages 1
Finished protocol to clipboard forwarder processed messages 1
Finished tcptls reader processed messages 0
Finished tcptls reader processed messages 1
Listening on local_addr=127.0.0.1:12903 protocol=tcp-tls multicast_addr=None certificate_serial=46:8d:f7:de:e1:27:f9:0d:06:c6:1f:ea:05:36:d8:2e:ad:69:09:35 certificate_dns=localhost
Listening on local_addr=[::1]:12903 protocol=tcp-tls multicast_addr=None certificate_serial=46:8d:f7:de:e1:27:f9:0d:06:c6:1f:ea:05:36:d8:2e:ad:69:09:35 certificate_dns=localhost
Loading certificates from configuration directory
Loading certificates from configuration directory
Message received data_size=102400 group=default message_type=file
New connection remote_addr=127.0.0.1:12904
Ntp server not provided
Received bytes 102400 in clipboard
Received stream expected_size=102444
Receiver connection closed: Connection expected data, but none was received
Sending warning alert CloseNotify
Starting clipboard writer
Tcp receive stream local_addr=127.0.0.1:12903 remote_addr=127.0.0.1:12904
Tcp receive stream local_addr=127.0.0.1:12903 remote_addr=127.0.0.1:12904
Validate data identity=127.0.0.1
Waiting for stream from remote_addr=127.0.0.1:12904
Waiting for stream from remote_addr=127.0.0.1:12904
Waiting to receive clipboard once
decided upon suite TLS13_AES_256_GCM_SHA384
